waspls/src/Wasp/Backend/ConcreteParser.hs

```diff
@@ -0,0 +1,98 @@
+module Wasp.Backend.ConcreteParser
+  ( -- * Concrete parser
+
+    -- | This module contains functions for converting lists of "Token"s into
+    -- concrete syntax trees. These trees represent any source file completely,
+    -- including whitespace and comments, even if the source file is invalid.
+    parseCST,
+    parseCSTExpression,
+  )
+where
+
+import Wasp.Backend.ConcreteParser.Internal
+import Wasp.Backend.ConcreteSyntax (SyntaxKind (..), SyntaxNode)
+import Wasp.Backend.ParseError
+import Wasp.Backend.Token (Token)
+import qualified Wasp.Backend.Token as T
+
+-- | Parse a list of tokens into a concrete syntax tree for an entire wasp file.
+--
+-- See "SyntaxNode" for a description of what a concrete syntax tree contains.
+parseCST :: [Token] -> ([ParseError], [SyntaxNode])
+parseCST tokens = parse tokens (root --> eof)
+
+-- | Parse a list of tokens into a concrete syntax tree for a wasp expression.
+-- This is mainly used for testing.
+--
+-- See "SyntaxNode" for a description of what a concrete syntax tree contains.
+parseCSTExpression :: [Token] -> ([ParseError], [SyntaxNode])
+parseCSTExpression tokens = parse tokens (expr --> eof)
+
+root :: GrammarRule
+root = group Program stmts
+
+stmts :: GrammarRule
+stmts = eof <> (stmt --> stmts)
+
+stmt :: GrammarRule
+stmt =
+  group Decl $
+    (T.Identifier `as` DeclType)
+      --> (T.Identifier `as` DeclName)
+      --> expr
+
+expr :: GrammarRule
+expr =
+  group Dict (listLike lcurly dictEntry comma rcurly)
+    <> group List (listLike lsquare expr comma rsquare)
+    <> group Quoter (lquote --> quotedText)
+    -- Note that we don't check number of tuple element here: this is left to
+    -- the next phase of parsing.
+    <> group Tuple (listLike lparen expr comma rparen)
+    <> int
+    <> double
+    <> string
+    <> (T.Identifier `as` Var)
+    <> kwTrue
+    <> kwFalse
+    <> extImport
+
+dictEntry :: GrammarRule
+dictEntry = group DictEntry $ (T.Identifier `as` DictKey) --> colon --> expr
+
+quotedText :: GrammarRule
+quotedText = eof <> rquote <> (quoted --> quotedText)
```

Comment:
Aha so there are many Quoted ones! Ok, I get it then.
I see that now we have:
```hs
...
  <|> Quoter <$$> (lquote --> quotedText)
...
quotedText :: GrammarRule
quotedText = rquote <|> (quoted --> quotedText)
```

which kind of feels incorrect, because it is not just quoted text, but also quote.

Maybe we could do it like this:
```hs
...
  <|> quoter
...
quoter :: GrammarRule
quoter = Quoter <$$> (lquote <> quoterTail)
  where quoterTail = rquote <|> (quoted <> quoterTail)
```

Reply:
Changed to the 2nd suggestion.